nvflare/app_common/executors/ham.py

```diff
@@ -176,76 +212,323 @@ def execute(self, task_name: str, shareable: Shareable, fl_ctx: FLContext, abort
-                # received min responses - remember the time at which this happened
-                min_received_time = current_time
+            if rc != ReturnCode.OK:
+                self.log_error(fl_ctx, f"error sending aggr result to parent {self._parent_name}: {rc}")
 
-            if current_time - min_received_time >= self.wait_time_after_min_resps_received:
-                # we have waited long enough after min responses received
+            parent_task_seq = reply.get_header(EdgeTaskHeaderKey.TASK_SEQ, task_info.seq)
+            if parent_task_seq != task_info.seq:
+                # this task is done
+                self.log_info(fl_ctx, f"task {task_info.seq} is done: parent moved to task {parent_task_seq}")
                 break
 
-            time.sleep(0.5)
+            # have I received all possible responses from my children?
+            if self._num_children > 0:
+                received, _ = self._pending_clients_status()
+                if received >= self._num_children:
+                    self.log_info(fl_ctx, f"task {task_info.seq} is done: all {received} child clients are done!")
+                    break
+
+            wrc = conditional_wait(
+                waiter=None,
+                timeout=aggr_interval + random.uniform(0.0, 0.5),
+                abort_signal=abort_signal,
+                condition_cb=self._check_task_done,
+            )
+            if wrc == WaiterRC.ABORTED:
+                return make_reply(ReturnCode.TASK_ABORTED)
+            elif wrc == WaiterRC.IS_SET:
+                break
 
-        # return aggregation result
         received, total = self._pending_clients_status()
-        self.log_info(fl_ctx, f"process done after {time.time() - start_time} secs: {received=} {total=}")
+        self.log_info(fl_ctx, f"task done after {time.time() - self._task_start_time} secs: {received=} {total=}")
 
         if self._process_error:
             self.log_error(fl_ctx, "there is process error")
             return make_reply(ReturnCode.EXECUTION_EXCEPTION)
 
-        if received == 0:
-            # nothing received! This maybe ok
-            self.log_warning(fl_ctx, "nothing received - timeout")
-            # return make_reply(ReturnCode.TIMEOUT)
+        # still anything to be aggregated?
+        return self._make_aggr_report(task_info, fl_ctx)
+
+    def _check_task_done(self):
+        if self._task_done:
+            # force the conditional wait to stop
+            return WaiterRC.IS_SET
+
+    def _make_aggr_report(self, task_info: TaskInfo, fl_ctx: FLContext):
+        task_data = task_info.task
+        cookie_jar = task_data.get_cookie_jar()
+        with self._aggr_lock:
+            has_aggr_data = True
+            report = self._do_aggr(fl_ctx)
+            if not report:
+                has_aggr_data = False
+                report = Shareable()
+
+            self.log_info(fl_ctx, f"making aggr report to parent for task {task_info.seq}: {has_aggr_data=}")
```

Comment:
This log generates too many lines. Maybe turn it into debug?

Reply:
Sure